Tests/Get-BMVariable.Tests.ps1

```diff
@@ -0,0 +1,381 @@
+    }
+
+    if( $ForApplicationGroup )
+    {
+        $optionalParams['ApplicationGroupName'] = $ForApplicationGroup
+    }
+
+    if( $ForEnvironment )
+    {
+        $optionalParams['EnvironmentName'] = $ForEnvironment
+    }
+
+    if( $ForServer )
+    {
+        $optionalParams['ServerName'] = $ForServer
+    }
+
+    if( $ForServerRole )
+    {
+        $optionalParams['ServerRoleName'] = $ForServerRole
+    }
+
+    $script:result = Get-BMVariable -Session $session @optionalParams
+}
+
+Describe 'Get-BMVariable.when there are no global variables' {
+    It 'should return nothing' {
+        Init
+        WhenGettingVariable
+        ThenNothingReturned
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting a global variable' {
+    It 'should return the variable' {
+        Init
+        GivenVariable 'Fubar' -WithValue 'Snafu'
+        GivenVariable 'Snafu' -WithValue 'Fubar'
+        WhenGettingVariable 'Fubar'
+        ThenVariableReturned @{ 'Fubar' = 'Snafu' }
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting a variable that doesn''t exist' {
+    It 'should fail' {
+        Init
+        WhenGettingVariable 'IDONOTEXIST!!!!!!' -ErrorAction SilentlyContinue
+        ThenNothingReturned
+        ThenError 'does\ not\ exist'
+    }
+}
+
+Describe 'Get-BMVariable.when ignoring when a variable that doesn''t exist' {
+    It 'should write no errors' {
+        Init
+        WhenGettingVariable 'IDONOTEXIST!!!!!!' -ErrorAction Ignore
+        ThenNothingReturned
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting a variable that doesn''t exist using wildcards' {
+    It 'should fail' {
+        Init
+        WhenGettingVariable 'IDONOTEXIST*'
+        ThenNothingReturned
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting all global variables' {
+    It 'should return all variables' {
+        Init
+        GivenVariable 'Fubar' -WithValue 'Snafu'
+        GivenVariable 'Snafu' -WithValue 'Fubar'
+        WhenGettingVariable
+        ThenVariableReturned @{ 'Fubar' = 'Snafu'; 'Snafu' = 'Fubar' }
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting all global variable values' {
+    It 'should return all values' {
+        Init
+        GivenVariable 'Fubar' -WithValue 'Snafu'
+        GivenVariable 'Snafu' -WithValue 'Fubar'
+        WhenGettingVariable -ValueOnly
+        ThenVariableValuesReturned @( 'Snafu', 'Fubar' )
+        ThenNoErrorWritten
+    }
+}
+
+Describe 'Get-BMVariable.when getting global variable''s value' {
+    It 'should return just the variable''s value' {
+        Init
+        GivenVariable 'Fubar' -WithValue 'Snafu'
+        GivenVariable 'Snafu' -WithValue 'Fubar'
+        WhenGettingVariable 'Snafu' -ValueOnly
+        ThenVariableValuesReturned 'Fubar'
+    }
+}
+
+Describe 'Get-BMVariable.when getting an application''s variables' {

```

Comment:
Any tests for the behavior if the given scope for a variable don't exist, such as the application, application group, server, server role?